### DISCUSSION TITLE
Adding support for configurable metadata extraction

### DISCUSSION THREAD

**Discussion by dastbe** [arxanas/git-branchless#424] [Ideas]
One thing I appreciate about superlog is that it extracts relevant information into the description. However, as-is it's not customizable without directly integrating those extractors into the source. While I suspect you might be amenable to adding support for some popular conventions with common tools, I have to imagine you're less keen to do the same for internal tools 😄 .

Given that, have you given any thought to supporting configurable node descriptors? In my mind, something that allows you to define a regex with a "well-known" capture group, along with optional additional configuration for formatting and styling.

**Comment by arxanas** [comment]
Hi @dastbe, I'm definitely open to this if you file a pull request! I was thinking about two approaches:

- Regex-only. Simple to configure but limited.
- Shell out to a custom command to get metadata for each commit. More complicated, but can render non-deterministic things like code review status.

Those two would also benefit from having a template language for the smartlog. [Jujutsu](https://github.com/martinvonz/jj) has a template language, which might also interest you.

**Comment by dastbe** [comment]
> Regex-only. Simple to configure but limited.

So the changes to make this work are straightforward, though wanted to check with you regarding configuration.

reading up on multivars in git config, it looks like it would be semantically reasonable to add a config key `branchless.core.descriptor_regex` that could be repeatedly defined in configuration, and then aggregate all defined keys and use those to extract each descriptor. Does that sound reasonable?

> Shell out to a custom command to get metadata for each commit. More complicated, but can render non-deterministic things like code review status.

This would be interesting and would help with some subset of templating (ex. matching `Discussed in #424` and returning `github.com/arxanas/git-branchless/discussions/424`), though I feel like you'd also want to support batching for use cases that are calling out to external services. Might be premature optimization. I expect you'd also want to have the same kind of multivar behavior described above.

> Those two would also benefit from having a template language for the smartlog. [Jujutsu](https://github.com/martinvonz/jj) has a template language, which might also interest you.

Also very interesting, will have to take a look at the functionality there.

**Reply by arxanas** [reply]
> it looks like it would be semantically reasonable to add a config key branchless.core.descriptor_regex that could be repeatedly defined in configuration, and then aggregate all defined keys and use those to extract each descriptor. Does that sound reasonable?

That sounds good to me. Let's use the config key `branchless.smartlog.descriptorRegex` instead.
